clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -241,15 +245,29 @@ public void run() {
         log.debug("Beginning shutdown of Kafka producer I/O thread, sending remaining records.");
 
         // okay we stopped accepting requests but there may still be
-        // requests in the accumulator or waiting for acknowledgment,
+        // requests in the transaction manager, accumulator or waiting for acknowledgment,
         // wait until these are completed.
-        while (!forceClose && (this.accumulator.hasUndrained() || this.client.inFlightRequestCount() > 0)) {
+        while (!forceClose && ((this.accumulator.hasUndrained() || this.client.inFlightRequestCount() > 0) || hasPendingTransactionalRequests())) {
             try {
                 run(time.milliseconds());
             } catch (Exception e) {
                 log.error("Uncaught error in kafka producer I/O thread: ", e);
             }
         }
+
+        // Abort the transaction if any commit or abort didn't went through the transaction manager's queue
+        while (!forceClose && transactionManager != null && transactionManager.hasOngoingTransaction()) {
+            if (!transactionManager.isCompleting()) {
+                log.debug("Aborting incomplete transaction due to shutdown");
+                transactionManager.beginAbort();
```

Comment:
Hmm.. I was looking into how the producer handles various operations after the call to close(). Unsurprisingly, there is a bunch of inconsistency. For send(), as far as I can tell, the logic looks like this:

1. Check whether the sender thread is closed and raise an error if it is
2. Refresh metadata if needed
3. Check if the accumulator is closed

To be honest, I don't know why we don't just check if the sender thread has begun shutdown, but perhaps that's a separate issue. For transactional requests, it seems we don't currently do any checking to see whether the producer has been shutdown. At least I couldn't find any checks in `commitTransaction()`. 

To make this solution work, I think we should have those checks. Basically the user should see some kind of error indicating that the producer has already begun shutdown and the operation is no longer allowed. As it is, they would probably see an illegal state error or something. This could be a simple check in `TransactionManager` if we add a `close()` method as suggested elsewhere.

Reply:
I think I'll make that change with checking if the producer is closed. `throwIfProducerClosed` is currently only used in doSend and as I see changing it to check the sender won't have very dramatic changes in behavior. Besides I could reuse it in the transactional methods.

Reply:
Regarding the second part I'll put in a close method in TransactionManager what you were referring to:
```
    synchronized void abortPendingTransactionalRequests() {
        pendingRequests.forEach(handler ->
                handler.fatalError(new KafkaException("The producer closed forcefully")));
    }
```
This would fail the requests. The name or the exception thrown might be debatable :). I would apply this in the forceClose part of the sender. So if aborting the transaction normally wasn't successful we just fail them on force close.